### PR TITLE
Add pass for adjusting conv2d input shape to parameters

### PR DESCRIPTION
**Pull request by Erik-Lundell**
While Pytorch accepts conv parameters such
that the sampling does not align perfectly with the input, TOSA has this as a requirement. This pass adds slice ops if the input shape needs adjustment.

Change-Id: I81ad6f172a946644092363eb0ace1d53979090d4

### CODE REVIEW

File: backends/arm/passes/size_adjust_conv2d_pass.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        args=args,
+        kwargs=kwargs or {},
+    )
+
+
+class SizeAdjustConv2DPass(ExportPass):
+    """
+    Adjust the convolution input size to match perfectly with the
+    weight size, padding, stride and dilation parameters.
+    This is done by inserting a slice op to remove the uneven end of the input.
```
Comment (digantdesai) [inline backends/arm/passes/size_adjust_conv2d_pass.py:69]: Does this align with the behavior of conv2d in PyTorch Eager - https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html

File: backends/arm/passes/size_adjust_conv2d_pass.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            if len(slice_args) == 0:
+                continue
+
+            with graph_module.graph.inserting_before(node):
+                last_node = cast(torch.fx.Node, input_node)
+                for args in slice_args:
+                    slice_node = graph.create_node(
+                        "call_function", self.slice_op, (last_node,) + args
+                    )
+                    if is_quant_node(last_node):
```
Comment (digantdesai) [inline backends/arm/passes/size_adjust_conv2d_pass.py:114]: leave comment here for what might the graph look like,
non quant,
```
input -> slice -> conv2d
```
quant,
```
...
```

**Review by digantdesai — APPROVED** [review]

File: backends/arm/operators/op_conv2d.py [hunk 3]
```diff
@@ -40,7 +40,7 @@ def adjust_pad_if_needed(self, input, weight, stride, pad, dilation):
 
         if mod_remainder > pad:
             raise RuntimeError(
-                f"ignoring input element is not currently supported, got a large stride {stride}"
+                "This case should be handled by the SizeAdjustConv2d pass, is it enabled?"
```
[the remark this answers is not in the document]
Comment (Erik-Lundell) [reply backends/arm/operators/op_conv2d.py:43]: This pass should always be enabled in the ArmPassManager so hopefully an end user should never encounter this error. The idea behind the error message is to point a developer to the pass if it is somehow removed in the future/ if there is a an edge case that the pass fails to handle.

File: backends/arm/passes/size_adjust_conv2d_pass.py (hunk 1 above)
Comment (Erik-Lundell) [reply backends/arm/passes/size_adjust_conv2d_pass.py:69]: Good question - I don't think that the documentation is detailed enough to determine that. What I could make out was to make sure that the shapes match but exactly how to truncate the padding/ data was unclear to me. Instead, I matched my implementation with our current Conv2d implementation which means always truncate "from the right". However, in some cases I guess you could save data by truncating the left-side padding first.

Do you have a suggestion for where to get more details about the PyTorch behaviour? Otherwise I guess you just need to add more tests than the two I added now - I created an internal ticket to investigate this more thoroughly.

File: backends/arm/passes/size_adjust_conv2d_pass.py (hunk 2 above)
Comment (Erik-Lundell) [reply backends/arm/passes/size_adjust_conv2d_pass.py:114]: Ack, I'll add this to the above mentioned ticket.

File: backends/arm/passes/size_adjust_conv2d_pass.py (hunk 1 above)
Comment (digantdesai) [reply backends/arm/passes/size_adjust_conv2d_pass.py:69]: > Do you have a suggestion for where to get more details about the PyTorch behaviour?

```
$ python
> import pytorch
> ...
``` 
😄
